In [2]:
! pip install -q pyomo


     |████████████████████████████████| 9.6 MB 5.3 MB/s 
     |████████████████████████████████| 49 kB 5.0 MB/s 


In [3]:
from pyomo.environ import *

In [4]:
import numpy as np

In [5]:
# create a model
model = ConcreteModel()

In [6]:
N = 20
M = 4

In [7]:
obj_coef_c = np.array([45, 48, 55, 65, 54, 37.5, 35, 52.5, 47, 28,82, 102, 150, 105, 95, 84, 88, 125, 70, 64])

In [8]:
constr_coef_A = ([[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],
                  [0.20, 0.25, 0.10, 0.30, 0.125, 0.35, 0.475, 0.10,0.20,0.50,0.20, 0.25, 0.10, 0.30, 0.125, 0.35, 0.475, 0.10,0.20,0.50 ],
                  [0.75, 0.65, 0.70, 0.60, 0.85, 0.525, 0.425, 0.40, 0.35, 0.40,0.75, 0.65, 0.70, 0.60, 0.85, 0.525, 0.425, 0.40, 0.35, 0.40],
                  [0.05, 0.10, 0.20, 0.10, 0.025, 0.125, 0.10, 0.50, 0.45, 0.10,0.05, 0.10, 0.20, 0.10, 0.025, 0.125, 0.10, 0.50, 0.45, 0.10]])
constr_rhs_b = ([425,116.875,244.375,63.75])

In [9]:
lower_bound = np.array([0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0])
upper_bound = np.array([15, 16, 9, 17.5, 12, 18.5, 2.5, 0.5, 9.5, 8,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf])

In [10]:
col_indices = np.arange(N)
row_indices = np.arange(M)
print(col_indices)
print(row_indices)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]
[0 1 2 3]


In [11]:
model.x = Var(col_indices)

In [12]:
model.constraints = ConstraintList()

In [13]:
for i in row_indices:
  model.constraints.add(sum(constr_coef_A[i][j]*model.x[j] for j in col_indices) >= constr_rhs_b[i])

In [14]:
for j in col_indices:
  model.x[j].setlb(lower_bound[j])
  model.x[j].setub(upper_bound[j])

In [15]:
model.objective = Objective(expr = sum(obj_coef_c[j]*model.x[j] for j in col_indices), sense=minimize)

In [16]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   20 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19}

1 Var Declarations
    x : Size=20, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  15.0 : False :  True :  Reals
          1 :     0 :  None :  16.0 : False :  True :  Reals
          2 :     0 :  None :   9.0 : False :  True :  Reals
          3 :     0 :  None :  17.5 : False :  True :  Reals
          4 :     0 :  None :  12.0 : False :  True :  Reals
          5 :     0 :  None :  18.5 : False :  True :  Reals
          6 :     0 :  None :   2.5 : False :  True :  Reals
          7 :     0 :  None :   0.5 : False :  True :  Reals
          8 :  

In [17]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 155680 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [18]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 28747.1875
  Upper bound: 28747.1875
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 21
  Number of nonzeros: 81
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.01233983039855957
# ----------------------------------------------------------
#   Solution Information
# -------------------------

In [19]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  28747.187499999993

Decision Variables
x[ 0 ] =  15.0
x[ 1 ] =  16.0
x[ 2 ] =  9.0
x[ 3 ] =  17.5
x[ 4 ] =  12.0
x[ 5 ] =  18.5
x[ 6 ] =  2.5
x[ 7 ] =  0.5
x[ 8 ] =  9.5
x[ 9 ] =  8.0
x[ 10 ] =  158.078125
x[ 11 ] =  0.0
x[ 12 ] =  0.0
x[ 13 ] =  0.0
x[ 14 ] =  0.0
x[ 15 ] =  0.0
x[ 16 ] =  0.0
x[ 17 ] =  0.0
x[ 18 ] =  74.046875
x[ 19 ] =  84.3749999999999

Constraints
constraints : Size=4
    Key : Lower   : Body               : Upper
      1 :   425.0 :  424.9999999999999 :  None
      2 : 116.875 : 116.87499999999994 :  None
      3 : 244.375 : 244.37499999999997 :  None
      4 :   63.75 :  63.74999999999999 :  None


Answer: 

LP model:=For production of 425 tons of metallic alloy product, Let Manil Metal Co. uses x[0],x[1],$\cdots$ x[9] tons of  alloys A-1,A-2 $\cdots $ A-10 from in-house stock and uses x[10],x[11],$\cdots$ x[19] tons of alloys A-1,A-2,$ \cdots  $ A-10 which they buy from market.now according to question if they use both the alloys then the cost function and constraints are following:  

Cost function =minimize  $45.0*x[0] + 48.0*x[1] + 55.0*x[2] + 65.0*x[3] + 54.0*x[4] + 37.5*x[5] + 35.0*x[6] + 52.5*x[7] + 47.0*x[8] + 28.0*x[9] + 82.0*x[10] + 102.0*x[11] + 150.0*x[12] + 105.0*x[13] + 95.0*x[14] + 84.0*x[15] + 88.0*x[16] + 125.0*x[17] + 70.0*x[18] + 64.0*x[19]$

Constraints:= 

$ \sum_{i=0}^{19} x_i \geq 425 \\  0.2*x[0] + 0.25*x[1] + 0.1*x[2] + 0.3*x[3] + 0.125*x[4] + 0.35*x[5] + 0.475*x[6] + 0.1*x[7] + 0.2*x[8] + 0.5*x[9] + 0.2*x[10] + 0.25*x[11] + 0.1*x[12] + 0.3*x[13] + 0.125*x[14] + 0.35*x[15] + 0.475*x[16] + 0.1*x[17] + 0.2*x[18] + 0.5*x[19] \geq 116.875\\  0.75*x[0] + 0.65*x[1] + 0.7*x[2] + 0.6*x[3] + 0.85*x[4] + 0.525*x[5] + 0.425*x[6] + 0.4*x[7] + 0.35*x[8] + 0.4*x[9] + 0.75*x[10] + 0.65*x[11] + 0.7*x[12] + 0.6*x[13] + 0.85*x[14] + 0.525*x[15] + 0.425*x[16] + 0.4*x[17] + 0.35*x[18] + 0.4*x[19] \geq 244.375 \\  0.05*x[0] + 0.1*x[1] + 0.2*x[2] + 0.1*x[3] + 0.025*x[4] + 0.125*x[5] + 0.1*x[6] + 0.5*x[7] + 0.45*x[8] + 0.1*x[9] + 0.05*x[10] + 0.1*x[11] + 0.2*x[12] + 0.1*x[13] + 0.025*x[14] + 0.125*x[15] + 0.1*x[16] + 0.5*x[17] + 0.45*x[18] + 0.1*x[19] \geq 63.75 $



Answer2:
 
 Now we get optimal cost:28747.187499999993


optimal quantaties of alloys:
x[ 0 ] =  15.0 \\
x[ 1 ] =  16.0 \\
x[ 2 ] =  9.0 \\
x[ 3 ] =  17.5 \\
x[ 4 ] =  12.0 \\
x[ 5 ] =  18.5 \\
x[ 6 ] =  2.5 \\
x[ 7 ] =  0.5 \\
x[ 8 ] =  9.5 \\
x[ 9 ] =  8.0 \\
x[ 10 ] =  158.078125 \\
x[ 11 ] =  0.0 \\
x[ 12 ] =  0.0 \\
x[ 13 ] =  0.0 \\
x[ 14 ] =  0.0 \\
x[ 15 ] =  0.0 \\
x[ 16 ] =  0.0 \\
x[ 17 ] =  0.0 \\
x[ 18 ] =  74.046875 \\
x[ 19 ] =  84.3749999999999

Answer3:

According to solution ,we get that alloy A-1 ,A-9 and A-10 is needed to buy.
And alloy A-1,A-9 and A-10 are three alloys which are high ranked to be bought(158.078125,74.046875,84.3749999999999 respectively).And A-2,A-3,A-4 are three alloys which are low ranked to be bought since according to answer there is no need to buy them from market.their in-house stock is sufficient to produce the 425 product.

Answer4:

According to solution,all the alloys are utilized.since all in-house stock of each alloy is used to produce the product.

Answer5:

According to solution , yes,there are some alloys which need not be aquired to market.as A-2,A-3,A-4,A-5,A-6,A-7,A-8 are these alloys not to be required to buy from market.